### PR TITLE
Update recipe_dictionary so contains works as expected.

### LINKED ISSUE #22170: You keep re-learning the same recipe when disassembling an item
Game version:
O.C-24382-g55c2ef5
Operating system:
OSX 10.11.6
Tiles or curses:
Curses
Mods active:
Disable NPC needs, Folding Parts pack, Vehicle Additions pack, Tanks and Other Vehicles, Disable Religious Texts, No Antique Firearms
### Expected behavior
If you know the recipe for an item (e.g. hotplate), you shouldn't get the check to learn the recipe when disassembling the item multiple times and consequently learning it anew multiple times. 
### Actual behavior
When crafting an item the recipe memorization works properly and recognizes which recipes are already memorizes and which not. This seems to be not working for recipe learning from disassembling items. Repeated disassembling will result in multiple messages that you can learn the recipe if you try again, or that you have successfully learned it (several times in a row sometimes). 
### Steps to reproduce the behavior
Disassemble any, for example, electronics item which recipe you already know. I've seen this so far with hotplates, soldering iron and many other.

### PR DESCRIPTION
**Pull request by laeaton2001**
Fixes #22170. The contains method in recipe_subset did not appear to be comparing recipe objects correctly. There is probably a much more efficient way of fixing this; still getting to grips with C++ & the code base.

### CODE REVIEW

**Comment by Coolthulhu** [comment]
`src/recipe_dictionary.h` is an astyled file meaning you need to run `make astyle` after changing it or manually adjust the file to match the styling.

File: src/recipe_dictionary.h [hunk 1]
```diff
@@ -89,7 +89,12 @@ class recipe_subset
 
         /** Check if the subset contains a recipe with the specified id. */
         bool contains( const recipe *r ) const {
-            return recipes.find( r ) != recipes.end();
+            for (const auto &elem : recipes) {
+                if (elem->ident() == r->ident()) {
+                    return true;
+                }
+            }
+            return false;
```
Comment (Coolthulhu) [inline src/recipe_dictionary.h]: Would be cleaner with the extension from `<functional>` (already included):
```
return std::any_if( recipes.begin(), recipes.end(), [r]( const recipe *elem ) {
    return elem->ident() == r->ident();
} );
```

**Comment by laeaton2001** [comment]
Thanks Coolthulhu, I knew there was a cleaner way of doing this.

**Comment by laeaton2001** [comment]
CI build failed due to not being able to get Clang 5 package.

**Pull request merged**